libs/sandbox-internal/src/main/kotlin/net/corda/sandbox/internal/SandboxServiceImpl.kt

```diff
@@ -50,8 +52,11 @@ internal class SandboxServiceImpl @Activate constructor(
         )
 
     /**
-     * We use concurrent hash maps everywhere (including sets) to allow concurrent access whilst keeping gets lock free
+     * We use concurrent hash maps everywhere (including sets) to allow concurrent access whilst keeping gets lock free.
+     * We also employ a lock to ensure writes are done to every map/set group atomically and we don't try to install
+     * and uninstall bundles from OSGi concurrently.
      */
+    val bundleLock = ReentrantLock()
```

Comment:
Would it not better to use:
```suggestion
    private val bundleLock = ReentrantReadWriteLock()
```
?

`areInSameSandbox` is the read-only operation and read aspect of the lock can be used to allow operations to progress in parallel.